README
Dive and movement analyses and plots			-*-outline-*-
====================================

Set of functions to analyse time-depth recorder and location data.

Read ?diveMove for a quick overview of the major functionality.  A
vignette is also available by doing vignette("diveMove").


* News
======

** Changes/New Features

*** Version 0.7-8:

        * Better arrangement of widgets in plotTD() and its methods.

*** Version 0.7-7:

    	* S4 Documentation bug fix.

*** Version 0.7-6:

    	* Fixes bug in plotTD() to work with stricter object class
          management in R devel.

*** Version 0.7-5:

	* Full examples for movement functions.

*** Version 0.7-4:

	* distSpeed() no longer uses 'id' argument.

	* Complete examples for location filtering functions.

	* createTDR() now provides useful default for 'concurrentData',
          and test for missing 'dtime'.  Completed example accordingly.

	* Improved show() method for 'TDRcalibrate' objects.

	* plotTD(), and hence the plotTDR() methods now accept an ellipsis
          argument to pass to par() for more flexibility drawing the
          plot.  They also now accept a new argument 'cex.pts' to set the
          relative size of points.

*** Version 0.7-3:

	* detPhase() and detDive() are now private functions (renamed with
          a preceding dot), and should only be used through
          calibrateDepth().

	* Rearranged documentation, reflecting the higher importance of
          calibrateDepth() and calibrateSpeed() functions.

	* distSpeed() no longer uses 'speed' and 'subset' arguments, and
          always returns a matrix.  All location filtering functions are
          now adapted to this change.

*** Version 0.7-2:

	* More examples for important functions.

*** Version 0.7-1:

	* 'TDRcalibrate' method for plot() now plots correctly if no dives
          were identified.
	  New argument 'interact' to allow for static plots, without
          'tcltk' interaction.

	* plot method replaced by 'plotTDR', to avoid unusual introduction
          of arguments in methods and simplify code.  This method is now
          based on the new function plotTD(), which replaces plotDive().
          This system now makes the relationships between functions much
          more clear.

	* Function getAct() renamed to rleActivity() to make it clearer
          that it calculates statistics from running series of readings
          with the same activity.

*** Version 0.7-0:

	* Arguments 'seaerr', 'landerr', and 'divethres' were renamed to
          'wet.thr', 'dry.thr', and 'dive.thr'.  Likewise, slots
          'land.threshold', 'sea.threshold', and 'dive.threshold' in class
          TDRcalibrate were renamed to 'dry.thr', 'wet.thr', and
          'dive.thr', respectively, for consistency/appropriateness.

	* The argument 'contours' to rqPlot() should now be a list with 2
          components.

*** Version 0.6-9:

	* New arguments for calibrateDepth(), offering improved control
	  over detection of descent, bottom, and ascent phases.  We now
	  use rates of vertical movement for detection, with a further
	  argument for defining what determines a bottom wiggle.

	* New arguments for plotDive() to decide whether to plot legend
          and to set limits for depth plot.

	* Documentation for formal classes and methods was split to make
          them easier to read.

	* attendance() has been renamed to timeBudget().


* Overview
==========

Dive analysis usually involves handling of large amounts of data, as new
instruments allow for frequent sampling of variables over long periods of
time.  The aim of this package is to make this process more efficient for
summarizing and extracting information gathered by time-depth recorders
(TDRs, hereafter).  The principal motivation for developing diveMove was
to provide more flexibility during the various stages of analysis than is
available in popular commercial software.  This is achieved by making the
results from intermediate calculations easily accessible, allowing the
user to make his/her own summaries beyond the default choices the package
provides.


* NOTE
======

Comments from users, especially bug reports, are very welcome:

Sebastian P. Luque <spluque@gmail.com>


ChangeLog
2014-02-20  Sebastian P. Luque <spluque@gmail.com>

	Releasing version 1.3.9

	* R/bouts.R, R/calibrate.R:
	  Declare global variables, as per post-submission message from
	  CRAN. Minor administrative change to satisfy the check procedure.

2014-01-30  Sebastian P. Luque <spluque@gmail.com>

	Releasing version 1.3.8

	* R/AllClass.R, R/AllMethod.R, R/calibrate.R, R/detDive.R,
	  R/detPhase.R, R/diveStats.R, R/oneDiveStats.R, R/plotTDR.R,
	  R/plotZOC.R, R/readTDR.R, R/stampDive.R, R/zoc.R:
	  Removed self triple colon references. They are not needed, and
	  are now raising NOTEs during checks with R-devel.

	* ChangeLog, DESCRIPTION:
	  New version.

2014-01-29  Sebastian P. Luque <spluque@gmail.com>

	* inst/doc, vignettes, vignettes/Makefile:
	  Moving vignettes to /vignettes directory.

2013-06-20  Sebastian P. Luque <spluque@gmail.com>

	* R/AllMethod.R:
	  Default concurrentData data.frame for createTDR() should have the
	  same number of rows as time stamps provided (thanks to Elizabeth
	  Masden for reporting).

2013-04-02  Sebastian P. Luque <spluque@gmail.com>

	* man/distSpeed.Rd:
	  Compare the entire matrix between the methods.

	* man/diveMove-package.Rd, man/dives.Rd, man/readLocs.Rd,
	  man/readTDR.Rd, man/sealLocs.Rd:
	  Streamline examples.

2013-04-01  Sebastian P. Luque <spluque@gmail.com>

	* man/dives.Rd:
	  Specify measuring interval.

	* tests/calibrations.R, tests/class-tests.R, tests/final-stats.R:
	  More meaningful tests.

2013-03-30  Sebastian P. Luque <spluque@gmail.com>

	* man/austFilter.Rd, man/distSpeed.Rd:
	  Better examples for distSpeed().

2013-03-29  Sebastian P. Luque <spluque@gmail.com>

	* R/austFilter.R, R/calibrate.R, R/distSpeed.R, R/zzz.R,
	  man/austFilter.Rd, man/distSpeed.Rd:
	  Implemented distance algorithms from geosphere package.

	* DESCRIPTION, NAMESPACE:
	  More efficient namespace set up.

2013-03-28  Sebastian P. Luque <spluque@gmail.com>

	* DESCRIPTION, NAMESPACE, R/plotTDR.R, R/zoc.R:
	  Use hsv() instead of depending on RColorBrewer package. Clean up
	  dependencies.

	* NEWS:
	  Do not use backtick to avoid fontification problems in ESS.

	Releasing version 1.3.6

	* NAMESPACE, NEWS, man/TDR-class.Rd:
	  Minor formatting of NEWS file to make it useable by news().

2013-01-31  Sebastian P. Luque <spluque@gmail.com>

	* R/AllMethod.R:
	  Formatting in show method.

	* R/AllMethod.R:
	  Previous commit wrongly changed "Dry threshold" rather than "Dive
	  threshold" in show method for TDRcalibrate. This fixes that and
	  specifies that the units are given by the instrument.

	* R/AllMethod.R:
	  Dive threshold is in m, not s. Thanks to Paolo Piedrahita.

2012-10-29  Sebastian P. Luque <spluque@gmail.com>

	* DESCRIPTION, R/detDive.R, R/stampDive.R,
	  man/plotDiveModel-methods.Rd:
	  New version.

	  Minor comment and documentation improvements.

2012-10-12  Sebastian P. Luque <spluque@gmail.com>

	* R/AllMethod.R:
	  Remove missing data from plotDiveModel() method for calculating
	  range of data, since there may be some depending on the values
	  used for 'wet.thr' and 'dry.thr' when 'interp.wet=TRUE' during
	  calibration.

2012-09-04  Sebastian P. Luque <spluque@gmail.com>

	* TODO:
	  Remove TODO item.

2012-08-15  Sebastian P. Luque <spluque@gmail.com>

	Releasing version 1.3.5

	* NEWS, R/calibrate.R, R/detPhase.R, man/calibrateDepth.Rd,
	  man/detPhase-internal.Rd:
	  New argument (logical) 'wet.cond' for .detPhase() and
	  calibrateDepth() to designate a condition for identifying wet/dry
	  conditions. Previously, this was done assuming that depth was
	  recorded only when wet (i.e. when a salt-water switch is
	  available and used accordingly). This assumption holds if new
	  argument 'wet.cond' is not supplied.

2012-07-30  Sebastian P. Luque <spluque@gmail.com>

	* man/calibrateDepth.Rd:
	  Minor wording change.

	* man/calibrateDepth.Rd:
	  Fixed factor level descriptions for dive phases (thanks to
	  Elizabeth Masden).

2012-05-28  Sebastian P. Luque <spluque@gmail.com>

	* DESCRIPTION, man/TDRcalibrate-class.Rd, man/bout-misc.Rd,
	  man/bouts2NLS.Rd, man/bouts3NLS.Rd, man/detDive-internal.Rd,
	  man/distSpeed.Rd, man/diveMove-internal.Rd, man/diveStats.Rd,
	  man/readTDR.Rd, man/sealLocs.Rd, man/timeBudget-methods.Rd:
	  Start new version and using links in help pages.

2012-05-04  Sebastian P. Luque <spluque@gmail.com>

	Releasing version 1.3.4

	* R/diveStats.R, man/diveMove-internal.Rd:
	  New argument for .derivStats() so that it extracts only requested
	  dives.  Thanks to Fabrizio Manco.

2012-04-13  Sebastian P. Luque <spluque@gmail.com>

	* inst/doc/diveMove.Rnw:
	  Increase resolution of vignette, but make sure we build package
	  with --compact-vignettes=gs+pdf.

2012-04-10  Sebastian P. Luque <spluque@gmail.com>

	* R/calibrate.R, man/diveMove-package.Rd, man/diveStats.Rd,
	  man/plotZOC-methods.Rd, man/bout-misc.Rd, man/bouts2MLE.Rd,
	  man/bouts2NLS.Rd, inst/doc/diveMove.Rnw:
	  Update to reflect trip to phase replacement for consistency.

	* R/stampDive.R:
	  Changed trip.no and trip.type to phase.no and activity for
	  consistency.

	* R/plotTDR.R:
	  Make resolution=100.

	* R/AllMethod.R:
	  Changed phaseno to phase.no for consistency.

	* man/calibrateDepth.Rd:
	  Corrected description for 'knot.factor'.

	* man/diveMove-internal.Rd:
	  Corrected elements for 'filter' method.

2012-04-05  Sebastian P. Luque <spluque@gmail.com>

	Releasing version 1.3.3

	* R/AllMethod.R, man/extractDive-methods.Rd:
	  Ignore duplicates in 'diveNo' argument.

	* DESCRIPTION, NEWS, R/AllClass.R, R/detDive.R:
	  Tighten validity of TDR objects so that time stamps are unique.

2012-03-28  Sebastian P. Luque <spluque@gmail.com>

	Releasing version 1.3.2

2012-03-27  Sebastian P. Luque <spluque@gmail.com>

	* DESCRIPTION, data/dives.csv, data/dives.csv.bz2,
	  data/sealLocs.csv, data/sealLocs.csv.bz2, inst/doc/diveMove.Rnw,
	  man/diveMove-package.Rd, man/dives.Rd, man/readLocs.Rd,

R/AllClass.R
## $Id: AllClass.R 600 2014-01-30 03:31:47Z sluque $

setClass("TDR",
         representation=representation(file="character", dtime="numeric",
             time="POSIXct", depth="numeric", concurrentData="data.frame"),
         prototype=prototype(concurrentData=data.frame()),
         validity=function(object) {
             if (length(object@time) != length(object@depth)) {
                 return("depth and time must have equal lengths")
             }
             time.diffs <- diff(unclass(object@time))
             if (any(time.diffs < 0)) {
                 return("time stamps must be in increasing order")
             }
             if (any(time.diffs == 0)) {
                 return("time stamps must not contain duplicate values")
             }
             ccDataN <- nrow(object@concurrentData)
             if (ccDataN > 0 && ccDataN != length(object@time)) {
                 mes <- paste("concurrentData must have the same number of rows",
                              "as there are time stamps")
                 return(mes)
             }
             if (!slot(object, "dtime")) return("dtime cannot be missing")
             return(TRUE)
         })

.speedNames <- c("velocity", "speed")
setClass("TDRspeed", contains="TDR",
         validity=function(object) {
             ccData <- object@concurrentData
             ccDataNames <- names(ccData)
             speedCol <- ccDataNames %in% .speedNames
             if (length(ccDataNames[speedCol]) != 1) {
                 return("speed is not available in concurrentData slot")
             } else if (!is.numeric(ccData[, speedCol])) {
                 return("speed must be of class numeric")
             }
             return(TRUE)
         })

setClass("TDRcalibrate",
         representation=representation(call="call", tdr="TDR",
           gross.activity="list", dive.activity="data.frame",
           dive.phases="factor", dive.models="list", dry.thr="numeric",
           wet.thr="numeric", dive.thr="numeric", speed.calib.coefs="numeric"),
         prototype=prototype(speed.calib.coefs=c(0, 1)),
         validity=function(object) {
             ndives <- max(object@dive.activity$dive.id, na.rm=TRUE)
             dml <- slot(object, "dive.models")
             if (length(slot(object, "dry.thr")) > 1) {
                 return("dry.thr must be a single number")
             }
             if (length(slot(object, "wet.thr")) > 1) {
                 return("wet.thr must be a single number")
             }
             if (length(slot(object, "dive.thr")) > 1) {
                 return("dive.thr must be a single number")
             }
             if (length(slot(object, "speed.calib.coefs")) != 2) {
                 return("speed.calib.coefs must be a length-2 vector")
             }
             if (length(dml) != ndives) {
                 return("All dives must have a corresponding dive model")
             }
             if (! all(sapply(dml, is, "diveModel"))) {
                 return("All elements of dive.models must be class diveModel")
             }
             return(TRUE)
         })

setOldClass("smooth.spline")
setClass("diveModel",
         representation=representation(label.matrix="matrix",
           dive.spline="smooth.spline", spline.deriv="list",
           descent.crit="numeric", ascent.crit="numeric",
           descent.crit.rate="numeric", ascent.crit.rate="numeric"),
         validity=function(object) {
             if (length(slot(object, "descent.crit")) > 1) {
                 return("descent.crit must be a single number")
             }
             if (length(slot(object, "ascent.crit")) > 1) {
                 return("ascent.crit must be a single number")
             }
             if (length(slot(object, "descent.crit.rate")) > 1) {
                 return("descent.crit.rate must be a single number")
             }
             if (length(slot(object, "ascent.crit.rate")) > 1) {
                 return("ascent.crit.rate must be a single number")
             }
             return(TRUE)
         })

setOldClass("nls")                      # For bout methods
